File: Apply Operations to Maximize Score/Apply Operations to Maximize Score.java
// Solution for Apply Operations to Maximize Score in JAVA

class Solution {
    private static final int MOD = (int) 1e9 + 7;
    
    private long findPower(long a, long b) {
        if (b == 0) return 1;
        long half = findPower(a, b / 2);
        long result = (half * half) % MOD;
        if (b % 2 == 1) {
            result = (result * a) % MOD;
        }
        return result;
    }
    
    private List<Integer> getPrimes(int limit) {
        boolean[] isPrime = new boolean[limit + 1];
        Arrays.fill(isPrime, true);
        List<Integer> primes = new ArrayList<>();
        for (int i = 2; i * i <= limit; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= limit; j += i) {
                    isPrime[j] = false;
                }
            }
        }
        for (int i = 2; i <= limit; i++) {
            if (isPrime[i]) {
                primes.add(i);
            }
        }
        return primes;
    }
    
    private int[] findPrimeScores(List<Integer> nums) {
        int n = nums.size();
        int[] primeScores = new int[n];
        int maxEle = Collections.max(nums);
        List<Integer> primes = getPrimes(maxEle);
        
        for (int i = 0; i < n; i++) {
            int num = nums.get(i);
            for (int prime : primes) {
                if (prime * prime > num) break;
                if (num % prime != 0) continue;
                primeScores[i]++;
                while (num % prime == 0) num /= prime;
            }
            if (num > 1) primeScores[i]++;
        }
        return primeScores;
    }
    
    private int[] findNextGreater(int[] primeScores) {
        int n = primeScores.length;
        int[] nextGreater = new int[n];
        Stack<Integer> stack = new Stack<>();
        
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && primeScores[stack.peek()] <= primeScores[i]) {
                stack.pop();
            }
            nextGreater[i] = stack.isEmpty() ? n : stack.peek();
            stack.push(i);
        }
        return nextGreater;
    }
    
    private int[] findPrevGreater(int[] primeScores) {
        int n = primeScores.length;
        int[] prevGreater = new int[n];
        Stack<Integer> stack = new Stack<>();
        
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && primeScores[stack.peek()] < primeScores[i]) {
                stack.pop();
            }
            prevGreater[i] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(i);
        }
        return prevGreater;
    }
    
    public int maximumScore(List<Integer> nums, int k) {
        int[] primeScores = findPrimeScores(nums);
        int[] nextGreater = findNextGreater(primeScores);
        int[] prevGreater = findPrevGreater(primeScores);
        
        int n = nums.size();
        long[] subarrays = new long[n];
        for (int i = 0; i < n; i++) {
            subarrays[i] = (long) (nextGreater[i] - i) * (i - prevGreater[i]);
        }
        
        List<int[]> sortedNums = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            sortedNums.add(new int[]{nums.get(i), i});
        }
        sortedNums.sort(Comparator.comparingInt(a -> a[0]));
        
        long score = 1;
        int idx = n - 1;
        while (k > 0) {
            int[] element = sortedNums.get(idx--);
            int num = element[0], i = element[1];
            long ops = Math.min(k, subarrays[i]);
            score = (score * findPower(num, ops)) % MOD;
            k -= ops;
        }
        return (int) score;
    }
}

File: Apply Operations to Maximize Score/Apply Operations to Maximize Score.py
# Solution for Apply Operations to Maximize Score in PY

class Solution:
    MOD = int(1e9 + 7)

    def findPower(self, a: int, b: int) -> int:
        if b == 0:
            return 1
        half = self.findPower(a, b // 2)
        result = (half * half) % self.MOD
        if b % 2 == 1:
            result = (result * a) % self.MOD
        return result

    def getPrimes(self, limit: int) -> List[int]:
        isPrime = [True] * (limit + 1)
        primes = []
        for i in range(2, int(limit ** 0.5) + 1):
            if isPrime[i]:
                for j in range(i * i, limit + 1, i):
                    isPrime[j] = False
        for i in range(2, limit + 1):
            if isPrime[i]:
                primes.append(i)
        return primes

    def findPrimeScores(self, nums: List[int]) -> List[int]:
        n = len(nums)
        primeScores = [0] * n
        maxEle = max(nums)
        primes = self.getPrimes(maxEle)
        
        for i in range(n):
            num = nums[i]
            for prime in primes:
                if prime * prime > num:
                    break
                if num % prime != 0:
                    continue
                primeScores[i] += 1
                while num % prime == 0:
                    num //= prime
            if num > 1:
                primeScores[i] += 1
        return primeScores

    def findNextGreater(self, primeScores: List[int]) -> List[int]:
        n = len(primeScores)
        nextGreater = [n] * n
        stack = []
        
        for i in range(n - 1, -1, -1):
            while stack and primeScores[stack[-1]] <= primeScores[i]:
                stack.pop()
            if stack:
                nextGreater[i] = stack[-1]
            stack.append(i)
        return nextGreater

    def findPrevGreater(self, primeScores: List[int]) -> List[int]:
        n = len(primeScores)
        prevGreater = [-1] * n
        stack = []
        
        for i in range(n):
            while stack and primeScores[stack[-1]] < primeScores[i]:
                stack.pop()
            if stack:
                prevGreater[i] = stack[-1]
            stack.append(i)
        return prevGreater

    def maximumScore(self, nums: List[int], k: int) -> int:
        primeScores = self.findPrimeScores(nums)
        nextGreater = self.findNextGreater(primeScores)
        prevGreater = self.findPrevGreater(primeScores)
        
        n = len(nums)
        subarrays = [0] * n
        for i in range(n):
            subarrays[i] = (nextGreater[i] - i) * (i - prevGreater[i])
        
        sortedNums = sorted([(nums[i], i) for i in range(n)])
        
        score = 1
        idx = n - 1
        while k > 0:
            num, i = sortedNums[idx]
            idx -= 1
            ops = min(k, subarrays[i])
            score = (score * self.findPower(num, ops)) % self.MOD
            k -= ops
        
        return score
